File: src/main/java/com/toy/shopwebmvc/domain/Item.java
package com.toy.shopwebmvc.domain;

import com.toy.shopwebmvc.constant.Category;
import com.toy.shopwebmvc.exception.CommonException;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import static com.toy.shopwebmvc.constant.ApiResponseCode.BAD_REQUEST;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Item extends BaseDomain {

    @Id
    @GeneratedValue
    @Column(name = "item_id")
    private Long id;

    private String name;
    private String description;
    private int price;
    private int quantity;

    @Enumerated(EnumType.STRING)
    private Category category;

    @Builder(builderMethodName = "create", toBuilder = true)
    public Item(String name, String description, int price, int quantity, Category category) {
        this.name = name;
        this.description = description;
        this.price = price;
        this.quantity = quantity;
        this.category = category;
    }

    public void removeQuantity(int quantity) {
        int differenceQuantity = this.quantity - quantity;

        if (differenceQuantity < 0) {
            throw new CommonException(BAD_REQUEST);
        }

        this.quantity = differenceQuantity;
    }

    public void addQuantity(int quantity) {
        this.quantity += quantity;
    }
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/Item.kt
package spring.web.kotlin.domain

import jakarta.persistence.*
import spring.web.kotlin.constant.Category

@Entity
class Item private constructor(
    var name: String,
    var description: String,
    var price: Int,
    var quantity: Int,

    @Enumerated(EnumType.STRING)
    var category: Category
) {
    @Id
    @GeneratedValue
    @Column(name = "item_id")
    var id: Long? = null

    companion object {
        fun create(name: String, description: String, price: Int, quantity: Int, category: Category): Item {
            return Item(name, description, price, quantity, category)
        }
    }
    fun update(name: String, description: String, price: Int, quantity: Int, category: Category) {
        this.name = name
        this.description = description
        this.price = price
        this.quantity = quantity
        this.category = category
    }
}